File: go/v3_analytics.go
package aegis

import (
	"context"
	"fmt"
	"net/url"

	"github.com/ai-aegis-yatav/aegis-sdk/go/internal"
)

// V3AnalyticsService — V3 analytics (explain, layer stats, attack clusters, baseline).
type V3AnalyticsService struct {
	t *internal.Transport
}

func v3AnalyticsQuery(params map[string]string) string {
	if len(params) == 0 {
		return ""
	}
	q := url.Values{}
	for k, v := range params {
		if v != "" {
			q.Set(k, v)
		}
	}
	if encoded := q.Encode(); encoded != "" {
		return "?" + encoded
	}
	return ""
}

func (s *V3AnalyticsService) Explain(ctx context.Context, judgmentID string) (map[string]any, error) {
	var resp map[string]any
	if err := s.t.Do(ctx, "GET", fmt.Sprintf("/v3/analytics/explain/%s", judgmentID), nil, &resp); err != nil {
		return nil, err
	}
	return resp, nil
}

func (s *V3AnalyticsService) LayerStats(ctx context.Context, params map[string]string) (map[string]any, error) {
	var resp map[string]any
	path := "/v3/analytics/layer-stats" + v3AnalyticsQuery(params)
	if err := s.t.Do(ctx, "GET", path, nil, &resp); err != nil {
		return nil, err
	}
	return resp, nil
}

func (s *V3AnalyticsService) AttackClusters(ctx context.Context, body map[string]any) (map[string]any, error) {
	var resp map[string]any
	if err := s.t.Do(ctx, "POST", "/v3/analytics/attack-clusters", body, &resp); err != nil {
		return nil, err
	}
	return resp, nil
}

func (s *V3AnalyticsService) Baseline(ctx context.Context, params map[string]string) (map[string]any, error) {
	var resp map[string]any
	path := "/v3/analytics/baseline" + v3AnalyticsQuery(params)
	if err := s.t.Do(ctx, "GET", path, nil, &resp); err != nil {
		return nil, err
	}
	return resp, nil
}

File: rust/src/resources/v3_analytics.rs
use std::sync::Arc;

use reqwest::Method;
use serde_json::Value;

use crate::error::Result;
use crate::transport::Transport;

pub struct V3Analytics {
    transport: Arc<Transport>,
}

fn build_query(params: &Value) -> String {
    let mut parts: Vec<String> = Vec::new();
    if let Some(obj) = params.as_object() {
        for (k, v) in obj {
            let value_str = match v {
                Value::String(s) => s.clone(),
                Value::Null => continue,
                other => other.to_string(),
            };
            parts.push(format!("{}={}", k, value_str));
        }
    }
    if parts.is_empty() {
        String::new()
    } else {
        format!("?{}", parts.join("&"))
    }
}

impl V3Analytics {
    pub(crate) fn new(transport: Arc<Transport>) -> Self {
        Self { transport }
    }

    pub async fn explain(&self, judgment_id: &str) -> Result<Value> {
        self.transport
            .request(
                Method::GET,
                &format!("/v3/analytics/explain/{}", judgment_id),
                None::<&()>,
            )
            .await
    }

    pub async fn layer_stats(&self, params: &Value) -> Result<Value> {
        let path = format!("/v3/analytics/layer-stats{}", build_query(params));
        self.transport
            .request(Method::GET, &path, None::<&()>)
            .await
    }

    pub async fn attack_clusters(&self, body: &Value) -> Result<Value> {
        self.transport
            .request(Method::POST, "/v3/analytics/attack-clusters", Some(body))
            .await
    }

    pub async fn baseline(&self, params: &Value) -> Result<Value> {
        let path = format!("/v3/analytics/baseline{}", build_query(params));
        self.transport
            .request(Method::GET, &path, None::<&()>)
            .await
    }
}
